File: matlab/droid.m
% change this to the IP shown in your VM
GAZEBO_VM_IP = "192.168.98.129";
rosinit(GAZEBO_VM_IP)

% subscribe to topics
camera = rossubscriber('/droid/camera/image_raw');
drive = rospublisher("/cmd_vel", "geometry_msgs/Twist");

% show livestream from the robot while controlling it
figure; axis image;
frame = imshow([]);

try
    while true
        % perceive
        image = readImage(receive(camera, 0.1));
        set(frame, 'CData', image); % update livestream

        % drive around in a circle
        command = rosmessage(drive);
        command.Angular.Z = 0.5;
        command.Linear.X = 0.1;
        
        % tell the robot to do the thing
        send(drive, command);
    end
catch
    % once we're done, shutdown
    rosshutdown()
end

File: python/droid.py
"starter drc droid example in python"

# use roslibpy (ROS over websockets) because the python packages are only available in linux.
from roslibpy import Ros, Topic, Message
import cv2

# change this to the IP shown in your VM
GAZEBO_VM_IP = '192.168.98.129'

client = Ros(host=GAZEBO_VM_IP, port=9090)
client.run()

# consume frames from the mjpeg stream because rosbridge (websocket) encodes the binary
# images as b64, because it wraps all messages in JSON, causing a huge performance hit.
camera = cv2.VideoCapture(f"http://{GAZEBO_VM_IP}:8080/stream?topic=/droid/camera/image_raw")

drive = Topic(client, '/cmd_vel', 'geometry_msgs/Twist')
drive.advertise()

# tell the robot to drive around in a circle forever
drive.publish(Message({
    'linear': {'x': 0.1},
    'angular': {'z': 0.5}
}))


try:
    while client.is_connected:
        _, image = camera.read()

        cv2.imshow('RobotCamera', image)
        cv2.waitKey(1)
finally:
    cv2.destroyAllWindows()
